File: src/widgets/kubernetes.zig
const std = @import("std");
const themes = @import("../themes/registry.zig");
const Theme = @import("../core/theme.zig").Theme;
const WidgetContext = @import("../core/widget.zig").WidgetContext;

fn getKubectlOutput(allocator: std.mem.Allocator, io: std.Io, argv: []const []const u8) !?[]u8 {
    const result = std.process.run(allocator, io, .{
        .argv = argv,
    }) catch return null;
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);

    if (result.term != .exited or result.term.exited != 0) {
        return null;
    }

    const trimmed = std.mem.trim(u8, result.stdout, " \n\r\t");
    if (trimmed.len == 0) {
        return null;
    }

    return try allocator.dupe(u8, trimmed);
}

fn contextContains(context: []const u8, needle: []const u8) bool {
    return std.ascii.indexOfIgnoreCase(context, needle) != null;
}

fn contextMatchesAny(context: []const u8, needles: []const []const u8) bool {
    for (needles) |needle| {
        if (contextContains(context, needle)) return true;
    }
    return false;
}

fn contextColor(theme: Theme, context: []const u8) []const u8 {
    if (contextMatchesAny(context, &.{ "prod", "production" })) return theme.danger;
    if (contextMatchesAny(context, &.{ "stage", "staging", "dev", "development" })) return theme.warning;
    return theme.info;
}

test "contextColor maps environment names semantically" {
    const theme = themes.hard;
    try std.testing.expectEqualStrings(theme.danger, contextColor(theme, "prod-cluster"));
    try std.testing.expectEqualStrings(theme.warning, contextColor(theme, "staging-cluster"));
    try std.testing.expectEqualStrings(theme.warning, contextColor(theme, "dev-cluster"));
    try std.testing.expectEqualStrings(theme.info, contextColor(theme, "sandbox"));
}

pub fn run(
    allocator: std.mem.Allocator,
    io: std.Io,
    theme_name: []const u8,
    transparent: bool,
    environ_map: *std.process.Environ.Map,
    writer: *std.Io.Writer,
) !void {
    var ctx_widget = try WidgetContext.init(allocator, io, environ_map, theme_name, transparent);
    defer ctx_widget.deinit();
    const theme = ctx_widget.theme;
    const reset = ctx_widget.reset;

    const context = getKubectlOutput(allocator, io, &.{
        "kubectl", "config", "current-context",
    }) catch return;
    defer if (context) |c| allocator.free(c);
    const ctx = context orelse return;

    const ns_opt = getKubectlOutput(allocator, io, &.{
        "kubectl", "config", "view", "--minify", "--output", "jsonpath={..namespace}",
    }) catch null;
    defer if (ns_opt) |ns| allocator.free(ns);
    const namespace = ns_opt orelse "default";

    const color = contextColor(theme, ctx);

    try writer.print("{s}#[fg={s},bg={s},bold]󱃾 {s}/{s}", .{
        reset,
        color,
        theme.background,
        ctx,
        namespace,
    });
}

File: src/widgets/kubernetes.rs
use std::process::Command;

use crate::core::cache::{cached_or_compute, scoped_cache_path};
use crate::core::widget::WidgetContext;
use crate::core::{Color, Theme};
use crate::tmux_renderer::ThemeHex;

// kubectl reads/validates the full kubeconfig on every invocation (and can
// hit the cluster API for some subcommands), so it is consistently the
// slowest of the subprocess-backed widgets. Two kubectl calls are needed per
// render (context, then namespace) — caching both together turns that into
// one cheap file read on most renders. Tradeoff: a `kubectl config
// use-context` / namespace switch can take up to this long to show up.
const KUBERNETES_CACHE_TTL_SECS: u64 = 5;

fn get_kubectl_output(argv: &[&str]) -> Option<String> {
    let output = Command::new("kubectl").args(argv).output().ok()?;
    if !output.status.success() {
        return None;
    }
    let trimmed = String::from_utf8_lossy(&output.stdout).trim().to_owned();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed)
    }
}

fn context_contains(context: &str, needle: &str) -> bool {
    context.to_lowercase().contains(&needle.to_lowercase())
}

fn context_matches_any(context: &str, needles: &[&str]) -> bool {
    needles
        .iter()
        .any(|needle| context_contains(context, needle))
}

fn context_color(theme: Theme, context: &str) -> Color {
    if context_matches_any(context, &["prod", "production"]) {
        theme.danger
    } else if context_matches_any(context, &["stage", "staging", "dev", "development"]) {
        theme.warning
    } else {
        theme.info
    }
}

/// Render the Kubernetes context widget.
/// Shows `󱃾 context/namespace` colored by environment (danger=prod, warning=stage/dev, info=other).
pub fn run(theme: Theme) -> String {
    let theme_hex = ThemeHex::from_theme(theme);
    run_with_theme_hex(theme, &theme_hex)
}

fn context_and_namespace() -> String {
    let context = get_kubectl_output(&["config", "current-context"]).unwrap_or_default();
    if context.is_empty() {
        // Cache the miss too, so a machine with no kubeconfig doesn't retry
        // kubectl on every render.
        return String::new();
    }

    let namespace = get_kubectl_output(&[
        "config",
        "view",
        "--minify",
        "--output",
        "jsonpath={..namespace}",
    ])
    .unwrap_or_else(|| String::from("default"));

    format!("{context}\n{namespace}")
}

pub(crate) fn run_with_theme_hex(theme: Theme, theme_hex: &ThemeHex) -> String {
    let ctx = WidgetContext::from_theme_hex(theme, theme_hex);
    let theme = ctx.theme;

    let cache_path = scoped_cache_path("kubernetes-context", None);
    let cached = cached_or_compute(&cache_path, KUBERNETES_CACHE_TTL_SECS, context_and_namespace);

    let mut parts = cached.splitn(2, '\n');
    let context = match parts.next() {
        Some(c) if !c.is_empty() => c,
        _ => return String::new(),
    };
    let namespace = parts.next().unwrap_or("default");

    let color = context_color(theme, context);

    format!(
        "{}#[fg={},bg={},bold]󱃾 {}/{}",
        ctx.reset,
        theme_hex.color(color),
        theme_hex.color(theme.surface),
        context,
        namespace,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::themes;

    #[test]
    fn context_color_maps_environment_names_semantically() {
        let theme = themes::hard::THEME;
        assert_eq!(theme.danger, context_color(theme, "prod-cluster"));
        assert_eq!(theme.warning, context_color(theme, "staging-cluster"));
        assert_eq!(theme.warning, context_color(theme, "dev-cluster"));
        assert_eq!(theme.info, context_color(theme, "sandbox"));
    }
}
